File: python-algorithm/leetcode/problem_92.py
"""92. Reverse Linked List II
https://leetcode.com/problems/reverse-linked-list-ii/

Given the head of a singly linked list and two integers left and right where
left <= right, reverse the nodes of the list from position left to position
right, and return the reversed list.

Example 1:

Input: head = [1,2,3,4,5], left = 2, right = 4
Output: [1,4,3,2,5]

Example 2:

Input: head = [5], left = 1, right = 1
Output: [5]

Constraints:

The number of nodes in the list is n.
1 <= n <= 500
-500 <= Node.val <= 500
1 <= left <= right <= n

Follow up: Could you do it in one pass?
"""
from common.list_node import ListNode


class Solution:
    def reverse_between(self, head: ListNode, left: int, right: int) -> ListNode:
        def reverse(node: ListNode) -> ListNode:
            prev = None
            while node:
                curr = node
                node = node.next
                curr.next = prev
                prev = curr
            return prev
            
        if not head or not head.next or left == right:
            return head
        dummy = ptr = ListNode(0, head)
        start = end = ptr
        i = 0
        while ptr:
            if i == left - 1:
                start = ptr
            if i == right:
                end = ptr.next
                ptr.next = None
            ptr = ptr.next
            i += 1
        old_start = start.next
        new_start = reverse(start.next)
        start.next = new_start
        old_start.next = end
        return dummy.next

File: java-algo/src/main/java/com/leetcode/Problem92.java
package com.leetcode;

import com.common.ListNode;

import java.util.ArrayList;
import java.util.List;

public class Problem92 {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if (head == null || head.next == null || left == right)
            return head;
        int pos = 0;
        ListNode dummy = new ListNode(0, head), p = dummy, start = null, end = null;
        List<ListNode> queue = new ArrayList<>();
        while (p != null) {
            if (pos == left - 1) {
                start = p;
            } else if (left <= pos && pos <= right) {
                queue.add(p);
                if (pos == right)
                    end = p.next;
            }
            p = p.next;
            pos++;
        }
        for (int i = queue.size() - 1; i > 0; i--) {
            queue.get(i).next = queue.get(i - 1);
        }
        queue.get(0).next = end;
        start.next = queue.get(queue.size() - 1);
        return dummy.next;
    }

    public ListNode reverseBetween2(ListNode head, int left, int right) {
        if (head == null || head.next == null || left == right)
            return head;
        int pos = 0;
        ListNode dummy = new ListNode(0, head), p = dummy, start = null, end = null;
        while (p != null) {
            if (pos == left - 1)
                start = p;
            if (pos == right)
                end = p.next;
            p = p.next;
            pos++;
        }
        helper(start, end);
        return dummy.next;
    }

    private void helper(ListNode leftNode, ListNode rightNode) {
        ListNode p1 = null, p2 = leftNode.next;
        while (p2 != rightNode) {
            ListNode nextNode = p2.next;
            p2.next = p1;
            p1 = p2;
            p2 = nextNode;
        }
        leftNode.next.next = rightNode;
        leftNode.next = p1;
    }
}
